Add j/k to scroll to next/prev tweet

paulmolluzzo opened the pull request:
This is a fix for #10.

I have to say, this feels super hacky to me, but it is working on the timeline and the notifications. There might be some voodoo in there and any feedback is welcome.

paulmolluzzo commented:
Ok, I made some improvements because the previous version didn't work at all. I didn't rebase so that you can see the progression of what happened, but the first commit actually doesn't work at all and the new version is working even as I go 20-40 tweets down in the timeline.

I am betting that there's going to be a lot of feedback on the code, and I'm cool with that. Please let me know if you find bugs or have code style feedback.

sindresorhus commented on extension/content.js:
```diff
@@ -54,6 +54,56 @@ function registerShortcuts(username) {
+	const navHeight = $('nav').clientHeight;
+	const tweets = document.querySelectorAll('._222QxFjc[role="row"]');
+	const currentTop = window.scrollY;
+	const keyCode = event.charCode;
```
Can you add a `TODO` comment here about using https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key when Chrome 51 is out?

sindresorhus commented on extension/content.js:
```diff
@@ -54,6 +54,56 @@ function registerShortcuts(username) {
+
+	Mousetrap.bind('j', event => {
+		scrollToTweet(event);
+	});
```
```js
Mousetrap.bind('j', scrollToTweet);
```

sindresorhus commented on extension/content.js:
```diff
@@ -54,6 +54,56 @@ function registerShortcuts(username) {
+	// shortcut for getting the total offset from the top of the document of a particular tweet
+	const totalOffset = tweetOffset => {
+		return tweetOffset + navHeight + tweetsContainerPadding;
+	};
```
```js
const totalOffset = tweetOffset => tweetOffset + navHeight + tweetsContainerPadding;
```

sindresorhus commented on extension/content.js:
```diff
@@ -54,6 +54,56 @@ function registerShortcuts(username) {
+	// takes a tweet's offsetTop and checks if it's below the navigation bar
+	const tweetIsBelowNav = offset => {
+		return totalOffset(offset) > currentTop;
+	};
```
Same as https://github.com/sindresorhus/refined-twitter/pull/19/files#r62426180

sindresorhus commented on extension/content.js:
```diff
@@ -54,6 +54,56 @@ function registerShortcuts(username) {
+
+	let scrollTarget = 0;
+
+	Array.from(tweets).some((tweet, index) => {
```
Can you add a `TODO` about removing `Array.from` when Chrome 51 is a target? With that version DOM collections are iterable \o/ https://mobile.twitter.com/chromiumdev/status/717736215433256960

sindresorhus commented:
This is awesome! Works almost perfectly. I only noticed one problem. On some tweets, I have to press `j` twice to get to the next tweet:

![untitled12](https://cloud.githubusercontent.com/assets/170270/15096051/eb0abae6-1511-11e6-8cba-3a7ffaf78175.gif)

sindresorhus commented:
Could you add the `j`/`k` shortcuts to the readme?

sindresorhus commented:
Maybe put it in a separate file so it can easily be ported to [Anatine](https://github.com/sindresorhus/anatine)?

paulmolluzzo commented:
@sindresorhus OK, I've addressed all the items you mentioned. Specifically:

* Fixed that occassional double-press scenario (see below)
* Added this to the README
* Streamlined the `totalOffset` and `tweetIsBelowNav` functions
* Added TODO items
* Moved this all to a new file
* Squashed all commits

Some notes:

The double-press issue is because of the way the math works out with all the `offset` grabbing and rounding, and in the end a magic number fixed it the best (I tried switching `ceil` to `floor` and changing `offset` to `getBoundingClientRect().top` and none of that worked for me). I'm not crazy about it, but it works. You can actually see the issue if you look - sometimes the top border of tweet perfect lines up, sometimes it's off by 1-5px.

I'm ignoring the first line of the new file to avoid the `no-unused-vars` error from xo.

sindresorhus commented:
Soooo good! :)

sindresorhus commented:
I ported it to Anatine: https://github.com/sindresorhus/anatine/commit/b70b860f1e86ab640b2e7e644e5c3efdded385ab :)

The focused tweet is not completely placed at the top though. Would you mind taking a look?

<img width="438" alt="napkin 09-05-16 2 25 04 pm" src="https://cloud.githubusercontent.com/assets/170270/15106197/d40708a0-15f1-11e6-85f5-78b00cad59ec.png">

paulmolluzzo commented:
@sindresorhus yup, I'll take a look.

paulmolluzzo commented:
@sindresorhus If it's working properly for refined-twitter, I suggest merging this PR and then handling the bug separately for anatine. It's possible that the solution needs to be customized for each repo based on the selectors/DOM we have in each.

sindresorhus commented:
@paulmolluzzo This is already merged.

hkdobrev commented:
I admit I still haven't installed the extension. But I've used Twitter.com for years and `j/k` navigation is provided by Twitter itself for a really long time. I'm sure it was available around 2009-2010 and it is still available now without the extension. So what's this about?

hkdobrev commented:
I'm sorry for commenting unprepared. So mobile.twitter.com doesn't have the keyboard shortcuts and this extension is all about mobile.twitter.com on desktop. This change makes perfect sense now. I'm probably not going to use it though. 😸

paulmolluzzo commented:
@hkdobrev it's already merged anyway :)

sindresorhus commented:
This is really awesome! Thanks for working on it :)

Any chance you could make it work when viewing a single tweet, like here: https://twitter.com/elonmusk/status/938972633416306690 ?

And can you move all the code into a new file in https://github.com/sindresorhus/refined-twitter/tree/master/source/features called `inline-instagram-photos.js` and just import it into content.js? It's better to have larger features self-contained.

sindresorhus commented:
> Any chance you could make it work when viewing a single tweet, like here: twitter.com/elonmusk/status/938972633416306690 ?

Oh, it actually works when you visit the link directly, but not when you click the tweet in the stream. I also noticed a race condition as sometimes it adds two of the same photo. Try refreshing and you might see it.

filipekiss commented:
Hey @sindresorhus, thanks for the feedback. 

Sure, I’ll make the adjustments you suggested and will take a look in how to prevent the race condition as soon as I get home! 

I’ll update the pull-request once I get things going!

filipekiss commented:
I've updated the code and I guess the race condition is now over.

About the single tweet, I'll make a separate commit for that and I'll update the PR soon enough.

filipekiss commented:
I've merged `upstream/master` onto my branch to get rid of the conflicts. I'll update the PR with the monkey patch soon.

filipekiss commented:
Well, I guess that's it! See if there's anything else that can be improved and I'll be happy to look into it.

sindresorhus commented:
It still duplicates the image, 3 times. Go to https://twitter.com/elonmusk/status/938972633416306690 and then refresh the page.

sindresorhus commented:
Can you also mention the feature in https://github.com/sindresorhus/refined-twitter#highlights ?

filipekiss commented:
@sindresorhus I'm looking into another way to detect and append the images, using `$.each` seems like it's not the best option. And as soon as I finish this, I'll update the README :)

filipekiss commented:
@sindresorhus Made the changes we discussed last week and some other minor upgrades. :)

sindresorhus commented:
Yay! Great work @filipekiss. This is going to be super useful for me :)
